controller/disk_controller.go

```diff
@@ -0,0 +1,541 @@
+			dc.enqueueDisk(disk)
+			err = nil
+		}
+	}()
+
+	// initialize the disk if necessary
+	if disk.Status.Conditions == nil {
+		disk.Status.Conditions = map[string]types.Condition{}
+	}
+	if disk.Status.ScheduledReplica == nil {
+		disk.Status.ScheduledReplica = map[string]int64{}
+	}
+	// when condition are not ready, the old storage data should be cleaned
+	disk.Status.StorageMaximum = 0
+	disk.Status.StorageAvailable = 0
+	disk.Status.StorageScheduled = 0
+	disk.Status.ScheduledReplica = map[string]int64{}
+
+	var diskFailureReason, diskFailureMessage string
+	defer func() {
+		if diskFailureReason != "" {
+			disk.Status.State = types.DiskStateDisconnected
+			disk.Status.Conditions = types.SetConditionAndRecord(disk.Status.Conditions,
+				types.DiskConditionTypeReady, types.ConditionStatusFalse, diskFailureReason, diskFailureMessage,
+				dc.eventRecorder, disk, v1.EventTypeWarning)
+			disk.Status.Conditions = types.SetConditionAndRecord(disk.Status.Conditions,
+				types.DiskConditionTypeSchedulable, types.ConditionStatusFalse,
+				string(types.DiskConditionReasonDiskNotReady),
+				"Disk is not ready",
+				dc.eventRecorder, disk, v1.EventTypeWarning)
+		}
+		if disk.Status.State == types.DiskStateConnected {
+			disk.Status.Conditions = types.SetConditionAndRecord(disk.Status.Conditions,
+				types.DiskConditionTypeReady, types.ConditionStatusTrue,
+				"", "",
+				dc.eventRecorder, disk, v1.EventTypeNormal)
+		}
+	}()
+
+	disk.Status.State = types.DiskStateConnected
+	if disk.Spec.NodeID == "" {
+		diskFailureReason = string(types.DiskConditionReasonNodeUnknown)
+		diskFailureMessage = "The Node ID is empty"
+		return nil
+	}
+	isNodeDown, err := dc.ds.IsNodeDownOrDeleted(disk.Spec.NodeID)
+	if err != nil {
+		return err
+	}
+	if isNodeDown {
+		diskFailureReason = string(types.DiskConditionReasonNodeUnknown)
+		diskFailureMessage = fmt.Sprintf("The connected node %v is down or deleted", disk.Spec.NodeID)
+		return nil
+	}
+	// Prevent the corner case:
+	// The corresponding ndoe is up but the ownership of the disk hasn't been transferred.
+	// In this case, Longhorn cannot verify UUID and FSID for the disk.
+	if disk.Spec.NodeID != dc.controllerID {
+		return nil
+	}
+
+	info, err := dc.getDiskInfoHandler(disk.Spec.Path)
+	if err != nil {
+		diskFailureReason = string(types.DiskConditionReasonNoDiskInfo)
+		diskFailureMessage = fmt.Sprintf("Get disk information, error: %v", err)
+		return nil
+	}
+
+	// Check disks in the same filesystem
+	if disk.Status.FSID == "" {
+		isDuplicate, err := dc.isFSIDDuplicatedWithExistingReadyDisk(disk, info.Fsid)
+		if err != nil {
+			return err
+		}
+		// Found multiple disks in the same Fsid
+		if isDuplicate {
+			diskFailureReason = string(types.DiskConditionReasonDiskFilesystemChanged)
+			diskFailureMessage = fmt.Sprintf("Disk has same file system ID %v as other disks", info.Fsid)
+			return nil
+		}
+		disk.Status.FSID = info.Fsid
+	} else {
+		if disk.Status.FSID != info.Fsid {
```

Comment:
disk FSID can change (https://github.com/longhorn/longhorn/blob/master/enhancements/20200331-replace-filesystem-id-key-in-disk-map.md#summary). That's why we introduced the disk UUID. We shouldn't use fsid to identify the disk.